tests/core/pthread/sync_to_async.cpp

```diff
@@ -0,0 +1,152 @@
+#include <assert.h>
+#include <emscripten.h>
+#include <pthread.h>
+
+#include <thread>
+#include <functional>
+#include <iostream>
+
+// Helper class for generic sync-to-async conversion. doWork() lets you provide
+// some code to be run, which is done on a helper thread. The thread can do
+// async operations, and it has a callback to indicate when it finishes working.
+// The calling thread mainwhile blocks as it waits.
+//
+// This can be useful if you are a location where blocking is possible (like a
+// thread, or when using PROXY_TO_PTHREAD), but you have code that is hard to
+// refactor to be async, but that requires some async operation (like waiting
+// for a JS event).
+class SyncToAsync {
+public:
+  using Callback = std::function<void()>;
```

Comment:
It might be cool to make this whole thing a template, so `Callback` (i.e. the resume function) could take an argument. Then that could be returned by `doWork`.
I guess that might make the implementation a little uglier (I'm not sure if you'd be able to use a lambda as you do currently on line 88?), but could make it more useful.

Reply:
I tinkered with this a little, but I'm not sure how this would help - maybe I'm not getting your idea?

This callback is just to notify the "runtime" that we have finished working. There isn't any argument we want to pass along there - the runtime wouldn't know what to do with any extra arguments?